File: solutions/lua/all-your-base/1/all-your-base.lua
local all_your_base = {}

local function reversed(xs)
	local sx = {}
	for i = #xs, 1, -1 do
		table.insert(sx, xs[i])
	end
	return sx
end

all_your_base.convert = function(from_digits, from_base)
	if from_base < 2 then
		error("invalid input base")
	end

	for _, v in ipairs(from_digits) do
		if v < 0 then
			error("negative digits are not allowed")
		end
		if v > (from_base - 1) then
			error("digit out of range")
		end
	end

	local res = 0
	for i = 0, #from_digits-1 do
		res = res + from_digits[i+1] * (from_base ^ (#from_digits - i - 1))
	end

	local o = {
		num = res
	}

	function o.to(base)
		if base < 2 then
			error("invalid output base")
		end
		local digits = {}
		local acc = o.num
		if acc == 0 then
			return { 0 }
		end
		while acc ~= 0 do
			table.insert(digits, acc % base)
			acc = (acc - (acc % base)) / base
		end
		return reversed(digits)
	end

	return o
end

return all_your_base

File: solutions/c/all-your-base/1/all_your_base.c
#include "all_your_base.h"

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ERROR_VALUE 0

static bool all_zeroes(int8_t* digits, size_t digits_length) {
	for (size_t i = 0; i < digits_length; i++)
		if (digits[i] != 0)
			return false;
	return true;
}

static size_t to_digits(int num, int from_base, int8_t* output_digits) {
	size_t digits_length = 0;
	for (int acc = num; acc != 0; acc /= from_base)
		output_digits[digits_length++] = acc % from_base;
	return digits_length;
}

static int from_digits(int8_t* digits, size_t digits_length, int from_base) {
	int res = 0;
	for (size_t i = 0; i < digits_length; i++)
		res = res*from_base + digits[i];
	return res;
}

static void reverse(int8_t* array, size_t length) {
	for (size_t i = 0; i < (length / 2); i++) {
		int8_t tmp = array[i];
		array[i] = array[length-i-1];
		array[length-i-1] = tmp;
	}
}

size_t rebase(int8_t* digits, int input_base, int output_base, int input_length) {
	if (input_base <= 1 || output_base <= 1)
		return ERROR_VALUE;
	if (input_length == 0)
		return 0;
	if (all_zeroes(digits, input_length)) {
		digits[0] = 0;
		return 1;
	}
	for (size_t i = 0; i < (size_t)input_length; i++) {
		int8_t d = digits[i];
		if (d >= input_base || d < 0)
			return ERROR_VALUE;
	}
	int num = from_digits(digits, input_length, input_base);
	size_t len = to_digits(num, output_base, digits);
	reverse(digits, len);
	return len;
}
